File: ts/src/execution/simple-agent-task-workflow.ts
import { LLMJudge } from "../judge/llm-judge.js";
import type { AgentTaskResult, LLMProvider } from "../types/index.js";
import type { JudgeInterface } from "../judge/delegated.js";
import type { RlmSessionRecord, RlmTaskConfig } from "../rlm/types.js";
import { runAgentTaskRlmSession } from "../rlm/agent-task.js";

export interface EvaluateSimpleAgentTaskOpts {
  taskPrompt: string;
  rubric: string;
  provider: LLMProvider;
  model: string;
  output: string;
  judgeOverride?: JudgeInterface;
  referenceContext?: string;
  requiredConcepts?: string[];
  calibrationExamples?: Array<Record<string, unknown>>;
  pinnedDimensions?: string[];
}

export async function evaluateSimpleAgentTaskOutput(
  opts: EvaluateSimpleAgentTaskOpts,
): Promise<AgentTaskResult> {
  const judge = opts.judgeOverride ?? new LLMJudge({
    provider: opts.provider,
    model: opts.model,
    rubric: opts.rubric,
  });
  const result = await judge.evaluate({
    taskPrompt: opts.taskPrompt,
    agentOutput: opts.output,
    referenceContext: opts.referenceContext,
    requiredConcepts: opts.requiredConcepts,
    calibrationExamples: opts.calibrationExamples,
    pinnedDimensions: opts.pinnedDimensions,
  });
  return {
    score: result.score,
    reasoning: result.reasoning,
    dimensionScores: result.dimensionScores,
    internalRetries: result.internalRetries ?? 0,
  };
}

export async function runSimpleAgentTaskRlm(opts: {
  provider: LLMProvider;
  model: string;
  config: RlmTaskConfig | null;
  phase: "generate" | "revise";
  taskPrompt: string;
  rubric: string;
  sessions: RlmSessionRecord[];
  revisionPrompt?: string;
  currentOutput?: string;
  judgeResult?: AgentTaskResult;
  referenceContext?: string;
  requiredConcepts?: string[];
}): Promise<string | null> {
  if (!opts.config) {
    return null;
  }
  const record = await runAgentTaskRlmSession({
    provider: opts.provider,
    model: opts.model,
    config: opts.config,
    phase: opts.phase,
    taskPrompt: opts.taskPrompt,
    rubric: opts.rubric,
    currentOutput: opts.currentOutput,
    judgeResult: opts.judgeResult,
    referenceContext: opts.referenceContext,
    requiredConcepts: opts.requiredConcepts,
    revisionPrompt: opts.revisionPrompt,
  });
  opts.sessions.push(record);
  const content = record.content.trim();
  return content.length > 0 ? content : null;
}

export async function generateSimpleAgentTaskOutput(opts: {
  provider: LLMProvider;
  model: string;
  taskPrompt: string;
  rubric: string;
  rlmConfig: RlmTaskConfig | null;
  rlmSessions: RlmSessionRecord[];
  referenceContext?: string;
  requiredConcepts?: string[];
}): Promise<string> {
  const rlmOutput = await runSimpleAgentTaskRlm({
    provider: opts.provider,
    model: opts.model,
    config: opts.rlmConfig,
    phase: "generate",
    taskPrompt: opts.taskPrompt,
    rubric: opts.rubric,
    sessions: opts.rlmSessions,
    referenceContext: opts.referenceContext,
    requiredConcepts: opts.requiredConcepts,
  });
  if (rlmOutput) {
    return rlmOutput;
  }

  const result = await opts.provider.complete({
    systemPrompt: "You are a skilled writer and analyst. Complete the task precisely.",
    userPrompt: buildSimpleAgentTaskUserPrompt({
      taskPrompt: opts.taskPrompt,
      referenceContext: opts.referenceContext,
      requiredConcepts: opts.requiredConcepts,
    }),
    model: opts.model,
  });
  return result.text;
}

export function buildSimpleAgentTaskUserPrompt(opts: {
  taskPrompt: string;
  referenceContext?: string;
  requiredConcepts?: string[];
}): string {
  const blocks = [
    opts.taskPrompt.trim(),
    buildReferenceContextBlock(opts.referenceContext),
    buildRequiredConceptsBlock(opts.requiredConcepts),
  ].filter((value) => value.length > 0);
  return blocks.join("\n\n");
}

export function buildSimpleAgentTaskRevisionPrompt(opts: {
  revisionPrompt?: string;
  output: string;
  judgeResult: AgentTaskResult;
  taskPrompt: string;
  referenceContext?: string;
  requiredConcepts?: string[];
}): string {
  const instruction = opts.revisionPrompt
    ?? "Revise the following output based on the judge's feedback. Maintain what works, fix what doesn't.";

  return [
    instruction,
    `## Original Output\n${opts.output}`,
    `## Judge Score: ${opts.judgeResult.score.toFixed(2)}`,
    `## Judge Feedback\n${opts.judgeResult.reasoning}`,
    buildReferenceContextBlock(opts.referenceContext),
    buildRequiredConceptsBlock(opts.requiredConcepts),
    `## Task\n${opts.taskPrompt}`,
    "Produce an improved version:",
  ].filter((value) => value.length > 0).join("\n\n");
}

export async function reviseSimpleAgentTaskOutput(opts: {
  provider: LLMProvider;
  model: string;
  taskPrompt: string;
  rubric: string;
  revisionPrompt?: string;
  output: string;
  judgeResult: AgentTaskResult;
  rlmConfig: RlmTaskConfig | null;
  rlmSessions: RlmSessionRecord[];
  referenceContext?: string;
  requiredConcepts?: string[];
}): Promise<string> {
  const rlmOutput = await runSimpleAgentTaskRlm({
    provider: opts.provider,
    model: opts.model,
    config: opts.rlmConfig,
    phase: "revise",
    taskPrompt: opts.taskPrompt,
    rubric: opts.rubric,
    sessions: opts.rlmSessions,
    revisionPrompt: opts.revisionPrompt,
    currentOutput: opts.output,
    judgeResult: opts.judgeResult,
    referenceContext: opts.referenceContext,
    requiredConcepts: opts.requiredConcepts,
  });
  if (rlmOutput) {
    return rlmOutput;
  }

  const result = await opts.provider.complete({
    systemPrompt:
      "You are revising content based on expert feedback. Improve the output. " +
      "IMPORTANT: Return ONLY the revised content. Do NOT include analysis, " +
      "explanations, headers like '## Revised Output', or self-assessment. " +
      "Just output the improved version directly.",
    userPrompt: buildSimpleAgentTaskRevisionPrompt({
      revisionPrompt: opts.revisionPrompt,
      output: opts.output,
      judgeResult: opts.judgeResult,
      taskPrompt: opts.taskPrompt,
      referenceContext: opts.referenceContext,
      requiredConcepts: opts.requiredConcepts,
    }),
    model: opts.model,
  });
  return result.text;
}

function buildReferenceContextBlock(referenceContext?: string): string {
  const trimmedReferenceContext = referenceContext?.trim();
  return trimmedReferenceContext ? `## Reference Context\n${trimmedReferenceContext}` : "";
}

function buildRequiredConceptsBlock(requiredConcepts?: string[]): string {
  const normalizedConcepts = requiredConcepts
    ?.map((concept) => concept.trim())
    .filter((concept) => concept.length > 0);
  return normalizedConcepts && normalizedConcepts.length > 0
    ? `## Required Concepts\n${normalizedConcepts.map((concept) => `- ${concept}`).join("\n")}`
    : "";
}

File: autocontext/src/autocontext/execution/simple_agent_task_workflow.py
"""Prompt helpers for simple queued agent tasks."""

from __future__ import annotations

from autocontext.providers.base import LLMProvider
from autocontext.scenarios.agent_task import AgentTaskResult


def generate_simple_agent_task_output(
    *,
    provider: LLMProvider,
    model: str,
    task_prompt: str,
    reference_context: str | None = None,
    required_concepts: list[str] | None = None,
) -> str:
    """Generate initial output for a simple queued task."""
    result = provider.complete(
        system_prompt="You are a skilled writer and analyst. Complete the task precisely.",
        user_prompt=build_simple_agent_task_user_prompt(
            task_prompt=task_prompt,
            reference_context=reference_context,
            required_concepts=required_concepts,
        ),
        model=model,
    )
    return result.text


def revise_simple_agent_task_output(
    *,
    provider: LLMProvider,
    model: str,
    task_prompt: str,
    output: str,
    judge_result: AgentTaskResult,
    revision_prompt: str | None = None,
    reference_context: str | None = None,
    required_concepts: list[str] | None = None,
    objective_feedback: str | None = None,
) -> str:
    """Revise output for a simple queued task."""
    result = provider.complete(
        system_prompt=(
            "You are revising content based on expert feedback. Improve the output. "
            "IMPORTANT: Return ONLY the revised content. Do NOT include analysis, "
            "explanations, headers like '## Revised Output', or self-assessment. "
            "Just output the improved version directly."
        ),
        user_prompt=build_simple_agent_task_revision_prompt(
            task_prompt=task_prompt,
            output=output,
            judge_result=judge_result,
            revision_prompt=revision_prompt,
            reference_context=reference_context,
            required_concepts=required_concepts,
            objective_feedback=objective_feedback,
        ),
        model=model,
    )
    return result.text


def build_simple_agent_task_user_prompt(
    *,
    task_prompt: str,
    reference_context: str | None = None,
    required_concepts: list[str] | None = None,
) -> str:
    """Build the direct-generation prompt for a simple queued task."""
    blocks = [
        task_prompt.strip(),
        _build_reference_context_block(reference_context),
        _build_required_concepts_block(required_concepts),
    ]
    return "\n\n".join(block for block in blocks if block)


def build_simple_agent_task_revision_prompt(
    *,
    task_prompt: str,
    output: str,
    judge_result: AgentTaskResult,
    revision_prompt: str | None = None,
    reference_context: str | None = None,
    required_concepts: list[str] | None = None,
    objective_feedback: str | None = None,
) -> str:
    """Build the revision prompt for a simple queued task."""
    instruction = revision_prompt or (
        "Revise the following output based on the judge's feedback. "
        "Maintain what works, fix what doesn't."
    )
    blocks = [
        instruction,
        f"## Original Output\n{output}",
        f"## Judge Score: {judge_result.score:.2f}",
        f"## Judge Feedback\n{judge_result.reasoning}",
        _build_objective_feedback_block(objective_feedback),
        _build_reference_context_block(reference_context),
        _build_required_concepts_block(required_concepts),
        f"## Task\n{task_prompt}",
        "Produce an improved version:",
    ]
    return "\n\n".join(block for block in blocks if block)


def _build_reference_context_block(reference_context: str | None) -> str:
    trimmed_reference_context = (reference_context or "").strip()
    if not trimmed_reference_context:
        return ""
    return f"## Reference Context\n{trimmed_reference_context}"


def _build_required_concepts_block(required_concepts: list[str] | None) -> str:
    normalized_concepts = [concept.strip() for concept in required_concepts or [] if concept.strip()]
    if not normalized_concepts:
        return ""
    concepts = "\n".join(f"- {concept}" for concept in normalized_concepts)
    return f"## Required Concepts\n{concepts}"


def _build_objective_feedback_block(objective_feedback: str | None) -> str:
    trimmed_objective_feedback = (objective_feedback or "").strip()
    if not trimmed_objective_feedback:
        return ""
    return f"## Objective Verification Feedback\n{trimmed_objective_feedback}"


__all__ = [
    "build_simple_agent_task_revision_prompt",
    "build_simple_agent_task_user_prompt",
    "generate_simple_agent_task_output",
    "revise_simple_agent_task_output",
]
